util/fee-estimator/src/tx_confirm_stat.rs

```diff
@@ -0,0 +1,318 @@
+use crate::FeeRate;
+use std::collections::BTreeMap;
+
+#[derive(Default, Debug, Clone)]
+struct BucketStat {
+    total_fee_rate: FeeRate,
+    txs_count: f64,
+    old_unconfirmed_txs: usize,
+}
+
+impl BucketStat {
+    // add a new fee rate to this bucket
+    fn new_fee_rate_sample(&mut self, fee_rate: FeeRate) {
+        self.txs_count += 1f64;
+        let total_fee_rate = self
+            .total_fee_rate
+            .as_u64()
+            .saturating_add(fee_rate.as_u64());
+        self.total_fee_rate = FeeRate::from_u64(total_fee_rate);
+    }
+
+    // get average fee rate from a bucket
+    fn avg_fee_rate(&self) -> FeeRate {
+        if self.txs_count > 0f64 {
+            FeeRate::from_u64((self.total_fee_rate.as_u64() / self.txs_count as u64) as u64)
+        } else {
+            FeeRate::zero()
+        }
+    }
+}
+
+/// TxConfirmStat is a struct to help to estimate txs fee rate,
+/// This struct record txs fee_rate and blocks that txs to be committed.
+///
+/// We start from track unconfirmed txs,
+/// When tx added to txpool, we increase the count of unconfirmed tx, we do opposite tx removed.
+/// When a tx get committed, put it into bucket by tx fee_rate and confirmed blocks,
+/// then decrease the count of unconfirmed txs.
+///
+/// So we get a group of samples which includes txs count, average fee rate and confirmed blocks, etc.
+/// For estimate, we loop through each bucket, calculate the confirmed txs rate, until meet the required_confirm_rate.
+#[derive(Clone)]
+pub struct TxConfirmStat {
+    /// per bucket stat
+    bucket_stats: Vec<BucketStat>,
+    /// bucket upper bound fee_rate => bucket index
+    fee_rate_to_bucket: BTreeMap<FeeRate, usize>,
+    /// confirm_blocks => bucket index => confirmed txs count
+    confirm_blocks_to_confirmed_txs: Vec<Vec<f64>>,
```

Comment:
Should it be `Vec<Vec<u64>>` since it's a counter?

Reply:
See decay

Reply:
We use decay to reduce the effect from old records, so it be an f64 can makes things easier.